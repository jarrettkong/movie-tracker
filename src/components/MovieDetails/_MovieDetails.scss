@import '../../variables.scss';

.MovieDetails {
	height: 100vh;
	width: calc(100vw - #{$nav-width});
	position: relative;
	color: white;
	z-index: 2;
	overflow: scroll;
}

.MovieDetails-content {
	padding: 30px;
	height: 100%;
	width: 100%;
	text-align: left;
	overflow: scroll;
}

.MovieDetails-backdrop {
	height: 100%;
	width: 100%;
	position: absolute;
	background-color: black;
	z-index: -1;
	& img {
		height: 100%;
		width: 100%;
		object-fit: cover;
		opacity: 0.15;
	}
}

h1 {
	margin: 0;
	font-size: 50px;
	font-weight: 600;
	& > span {
		font-weight: 300;
		font-size: 30px;
		& span {
			font-size: 16px;
		}
	}
}

.MovieDetails-genres {
	margin: 15px 0;
}

.MovieDetails-genre {
	padding: 5px;
	border-radius: 5px;
	border: 1px solid white;
	display: inline-block;
	margin: 0 5px;
	&:first-child {
		margin-left: 0;
	}
	&:last-child {
		margin-right: 0;
	}
}

.MovieDetails-main {
	border-right: 1px solid white;
	display: grid;
	grid-template-columns: 342px auto;
}

.MovieDetails-metadata {
	padding: 0 50px;
	line-height: 1.4;
	& .metadata-header {
		margin: 0 0 5px 0;
		font-weight: 600;
		font-size: 20px;
	}
	& .metadata-info {
		margin-top: 0;
	}
}

.MovieDetails-grid {
	display: grid;
	grid-template-columns: 4fr 1fr;
	grid-gap: 25px;
}

.MovieDetails-similar {
	& h2 {
		margin: 0;
	}
}
