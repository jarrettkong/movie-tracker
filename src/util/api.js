export const fetchMovies = async path => {
  const res = await fetch(
    `${process.env.REACT_APP_BASE_URL}/movie/${path}?api_key=${process.env.REACT_APP_API_KEY}`
  )
  if (!res.ok) {
    throw new Error('Failed to fetch movie data.');
  } 
  return await res.json();
}

export const fetchUser = async (body, dir) => {
  const res = await fetch(`http://localhost:3000/api/users/${dir}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  });
  if (!res.ok) {
    throw new Error('Failed to post user data.');
  } 
  return await res.json();
}