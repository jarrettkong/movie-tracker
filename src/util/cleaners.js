export const cleanMovies = (movies, configUrl) => {
	return movies.map(movie => {
		return {
			poster: movie.poster_path ? `${process.env.REACT_APP_BASE_IMAGE_URL}w300${movie.poster_path}` : null,
			id: movie.id,
			title: movie.title
			// releaseDate: movie.release_date
		};
	});
};

export const cleanMovieDetails = (movie, configUrl) => {
	return {
		poster: movie.poster_path ? `${process.env.REACT_APP_BASE_IMAGE_URL}w500${movie.poster_path}` : null,
		backdrop: movie.backdrop_path ? `${process.env.REACT_APP_BASE_IMAGE_URL}w1280${movie.backdrop_path}` : null,
		genres: movie.genres.map(genre => genre.name).join(', '),
		id: movie.id,
		rating: movie.vote_average,
		title: movie.original_title,
		language: movie.original_language,
		releaseDate: movie.release_date,
		description: movie.overview,
		runtime: movie.runtime,
		popularity: movie.popularity,
		productionCompanies: movie.production_companies.map(c => c.name).join(', '),
		boxOffice: movie.revenue,
		cast: '/credits',
		recommendatoins: '/recommendations'
	};
};
