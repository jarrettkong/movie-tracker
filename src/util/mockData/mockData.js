export const mockMovie = {
	adult: false,
	backdrop_path: '/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg',
	belongs_to_collection: {
		id: 86311,
		name: 'The Avengers Collection',
		poster_path: '/qJawKUQcIBha507UahUlX0keOT7.jpg',
		backdrop_path: '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'
	},
	budget: 356000000,
	genres: [
		{
			id: 12,
			name: 'Adventure'
		},
		{
			id: 878,
			name: 'Science Fiction'
		},
		{
			id: 28,
			name: 'Action'
		}
	],
	homepage: 'https://www.marvel.com/movies/avengers-endgame',
	id: 299534,
	imdb_id: 'tt4154796',
	original_language: 'en',
	original_title: 'Avengers: Endgame',
	overview:
		"After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store.",
	popularity: 335.962,
	poster_path: '/or06FN3Dka5tukK1e9sl16pB3iy.jpg',
	production_companies: [
		{
			id: 420,
			logo_path: '/hUzeosd33nzE5MCNsZxCGEKTXaQ.png',
			name: 'Marvel Studios',
			origin_country: 'US'
		}
	],
	production_countries: [
		{
			iso_3166_1: 'US',
			name: 'United States of America'
		}
	],
	release_date: '2019-04-24',
	revenue: 2489617092,
	runtime: 181,
	spoken_languages: [
		{
			iso_639_1: 'en',
			name: 'English'
		}
	],
	status: 'Released',
	tagline: 'Part of the journey is the end.',
	title: 'Avengers: Endgame',
	video: false,
	vote_average: 8.5,
	vote_count: 5198
};

export const mockUser = {
	id: 1,
	name: 'Bob',
	email: 'bob@bob.com',
	password: 'password'
};

export const mockPost = {
	method: 'POST',
	headers: { 'Content-Type': 'application/json' },
	body: JSON.stringify({ email: mockUser.email, password: mockUser.password })
};

export const mockPath = 'popular';

export const mockMovies = [
	{
		id: 456740,
		poster: 'https://image.tmdb.org/t/p/w300/bk8LyaMqUtaQ9hUShuvFznQYQKR.jpg',
		title: 'Hellboy'
	},
	{
		id: 447404,
		poster: 'https://image.tmdb.org/t/p/w300/wgQ7APnFpf1TuviKHXeEe3KnsTV.jpg',
		title: 'Pok√©mon Detective Pikachu'
	}
];

export const mockState = {
	trendingMovies: mockMovies,
	topRatedMovies: mockMovies,
	nowPlayingMovies: mockMovies,
	upcomingMovies: mockMovies,
	user: mockUser
};
