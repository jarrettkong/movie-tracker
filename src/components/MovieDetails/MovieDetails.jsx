import React, { Component } from 'react';
import { cleanMovieDetails } from '../../util/Cleaners';
import './_MovieDetails.scss';

class MovieDetails extends Component {
	state = {
		details: {},
		loading: false
	};

	componentDidMount() {
		const { REACT_APP_BASE_URL, REACT_APP_API_KEY } = process.env;
		this.setState({ loading: true }, async () => {
			try {
				const res = await fetch(`${REACT_APP_BASE_URL}/movie/${this.props.id}?api_key=${REACT_APP_API_KEY}`);
				const movie = await res.json();
				const details = cleanMovieDetails(movie);
				this.setState({ details, loading: false });
			} catch (err) {
				console.log(err);
			}
		});
	}

	render() {
		const {
			title,
			releaseDate,
			runtime,
			poster,
			backdrop,
			genres,
			// id,
			rating,
			language,
			description,
			popularity,
			production,
			boxOffice
		} = this.state.details;

		return (
			<section className="MovieDetails">
				{this.state.loading ? (
					<h1>Loading...</h1>
				) : (
					<div>
						<div className="MovieDetails-backdrop">
							<img src={backdrop} alt={`${title} backdrop`} />
						</div>
						<section className="MovieDetails-content">
							<header>
								<h1>
									{title}{' '}
									<span className="MovieDetails-release-date">
										({releaseDate}) <span>{language}</span>
									</span>
								</h1>
								<div className="MovieDetails-genres">
									{genres && genres.map(g => <p className="MovieDetails-genre">{g}</p>)}
								</div>
							</header>
							<div className="MovieDetails-grid">
								<main className="MovieDetails-main">
									<img src={poster} alt={`${title} poster`} />
									<div className="MovieDetails-metadata">
										<p className="metadata-header">Rating:</p>
										<p className="metadata-info">{rating}</p>
										<p className="metadata-header">Runtime:</p>
										<p className="metadata-info">{runtime} minutes</p>
										<p className="metadata-header">Description:</p>
										<p className="metadata-info">{description}</p>
										<p className="metadata-header">Production Companies:</p>
										<p className="metadata-info">{production}</p>
										<p className="metadata-header">Popularity:</p>
										<p className="metadata-info">{popularity}</p>
										<p className="metadata-header">Box Office Gross: </p>
										<p className="metadata-info">${boxOffice}</p>
									</div>
								</main>
								<section className="MovieDetails-similar">
									<h2>Similar Movies</h2>
								</section>
							</div>
						</section>
					</div>
				)}
			</section>
		);
	}
}

export default MovieDetails;
