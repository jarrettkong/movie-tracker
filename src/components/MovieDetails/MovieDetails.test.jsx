import React from 'react';
import { MovieDetails } from './MovieDetails';
import { shallow } from 'enzyme';
import * as Api from '../../util/api';
import * as Cleaners from '../../util/cleaners';
import { mockMovie, mockMovies, mockUser, mockMovieDetails } from '../../util/mockData/mockData';

jest.mock('../../util/api');
jest.mock('../../util/cleaners');

describe('MovieDetails', () => {
	let wrapper, mockAddMovieDetails;

	beforeEach(() => {
		window.scrollTo = jest.fn();
		mockAddMovieDetails = jest.fn();
		wrapper = shallow(
			<MovieDetails
				id={mockMovie.id}
				key={mockMovie.id}
				movies={mockMovies}
				favorites={mockMovies}
				user={mockUser}
				movieDetails={mockMovieDetails}
				addMovieDetails={mockAddMovieDetails}
			/>
		);
	});

	it('Should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('Should match the snapshot with the Loader', () => {
		wrapper.setState({ loading: true });
		expect(wrapper).toMatchSnapshot();
	});

	it('Should match the snapshot with the empty favorite icon', () => {
		wrapper = shallow(
			<MovieDetails
				id={mockMovie.id}
				key={mockMovie.id}
				movies={mockMovies}
				favorites={[]}
				user={mockUser}
				movieDetails={mockMovieDetails}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should scroll to the top on update', () => {
		wrapper.instance().componentDidUpdate();
		expect(window.scrollTo).toHaveBeenCalledWith(0, 0);
	});

	describe('componentDidMount', () => {
		it('should set state if it finds details', () => {
			jest.spyOn(wrapper.instance(), 'getMovieDetails');
			wrapper.instance().componentDidMount();
			expect(wrapper.state('details')).toEqual({ ...mockMovie, movie_id: mockMovie.id });
			expect(wrapper.instance().getMovieDetails).toHaveBeenCalledTimes(0);
		});

		it('should call getMovieDetails if there are no details', () => {
			wrapper = shallow(
				<MovieDetails
					id={mockMovie.id}
					key={mockMovie.id}
					movies={mockMovies}
					favorites={mockMovies}
					user={mockUser}
					movieDetails={[]}
				/>
			);
			jest.spyOn(wrapper.instance(), 'getMovieDetails');
			wrapper.instance().componentDidMount();
			expect(wrapper.instance().getMovieDetails).toHaveBeenCalled();
		});
	});

	describe('deleteFavorite', () => {
		it('should call deleteFavorite with the body', async () => {
			const body = {
				user_id: mockUser.id,
				movie_id: mockMovie.id
			};
			await wrapper.instance().deleteFavorite();
			expect(Api.deleteFavorite).toHaveBeenCalledWith(body);
		});

		it.skip('should set state to the error if it fails', async () => {
			wrapper = shallow(
				<MovieDetails
					id={mockMovie.id}
					key={mockMovie.id}
					movies={mockMovies}
					favorites={mockMovies}
					user={mockUser}
					movieDetails={[]}
				/>
			);
			await wrapper.instance().deleteFavorite();
			expect(wrapper.state('error')).toEqual('Could not delete favorite');
		});
	});

	describe('addFavorite', () => {
		it('should call addFavorite with the body', async () => {
			const body = {
				movie_id: mockMovie.id,
				user_id: mockUser.id,
				poster_path: mockMovie.poster_path,
				release_date: mockMovie.releaseDate,
				vote_average: mockMovie.rating,
				title: mockMovie.title,
				overview: mockMovie.description
			};
			await wrapper.instance().addFavorite();
			expect(Api.addFavorite).toHaveBeenCalledWith(body);
		});

		it.skip('should set state to the error if it fails', async () => {
			wrapper = shallow(
				<MovieDetails
					id={mockMovie.id}
					key={mockMovie.id}
					movies={mockMovies}
					favorites={mockMovies}
					user={{ ...mockUser, id: 111 }}
					movieDetails={[]}
				/>
			);
			await wrapper.instance().deleteFavorite();
			expect(wrapper.state('error')).toEqual('Could not delete favorite');
		});
	});

	describe('getMovieDetails', () => {
		beforeEach(() => {
			window.fetch = jest.fn().mockImplementation(() => {
				return Promise.resolve({
					ok: true,
					json: () => Promise.resolve(mockMovie)
				});
			});
		});

		it('should call fetch with the correct params', async () => {
			const { REACT_APP_BASE_URL, REACT_APP_API_KEY } = process.env;
			const url = `${REACT_APP_BASE_URL}/movie/${mockMovie.id}?api_key=${REACT_APP_API_KEY}`;
			await wrapper.instance().getMovieDetails();
			expect(window.fetch).toHaveBeenCalledWith(url);
		});

		it('should call cleanMovieDetails', async () => {
			await wrapper.instance().getMovieDetails();
			expect(Cleaners.cleanMovieDetails).toHaveBeenCalled();
		});

		it.skip('should call addMovieDetails', async () => {
			await wrapper.instance().getMovieDetails();
			expect(mockAddMovieDetails).toHaveBeenCalled();
		});

		it('should set state with the details', async () => {
			await wrapper.instance().getMovieDetails();
			expect(wrapper.state('details')).toEqual({ ...mockMovie, movie_id: mockMovie.id });
		});
	});

	describe('handleClick', () => {
		it('should call deleteFavorite', async () => {
			await wrapper.instance().handleClick();
			expect(Api.deleteFavorite).toHaveBeenCalled();
		});

		it.skip('should call addFavorite', async () => {
			wrapper = shallow(
				<MovieDetails
					id={mockMovie.id}
					key={mockMovie.id}
					movies={mockMovies}
					favorites={[]}
					user={{ ...mockUser, id: 111 }}
					movieDetails={[]}
				/>
			);
			await wrapper.instance().handleClick();
			expect(Api.addFavorite).toHaveBeenCalled();
		});

		it.skip('should call addFavorites', async () => {
			const mockAddFavorites = jest.fn();
			wrapper = shallow(
				<MovieDetails
					id={mockMovie.id}
					key={mockMovie.id}
					movies={mockMovies}
					favorites={[]}
					user={{ ...mockUser, movie_id: 111 }}
					movieDetails={[]}
					addFavorites={mockAddFavorites}
				/>
			);
			await wrapper.instance().getMovieDetails();
			expect(mockAddFavorites).toHaveBeenCalled();
		});
	});
});
