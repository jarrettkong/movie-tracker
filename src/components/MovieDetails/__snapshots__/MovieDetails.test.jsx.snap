// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`MovieDetails Should match the snapshot with the Loader 1`] = `ShallowWrapper {}`;

exports[`MovieDetails Should match the snapshot with the empty favorite icon 1`] = `ShallowWrapper {}`;
